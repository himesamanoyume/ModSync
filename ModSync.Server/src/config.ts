import path from "node:path";
import type { PreSptModLoader } from "@spt/loaders/PreSptModLoader";
import type { JsonUtil } from "@spt/utils/JsonUtil";
import type { VFS } from "@spt/utils/VFS";
import { glob } from "./utility/glob";
import { unixPath } from "./utility/misc";
import type { ILogger } from "@spt/models/spt/utils/ILogger";

export type SyncPath = {
	path: string;
	enabled?: boolean;
	enforced?: boolean;
	silent?: boolean;
	restartRequired?: boolean;
};

type RawConfig = {
	syncPaths: (string | SyncPath)[];
	exclusions: string[];
};

const DEFAULT_CONFIG = `{
	"syncPaths": [
		"BepInEx/plugins",
		"BepInEx/patchers",
		"BepInEx/config",
		{
			"enabled": false,
			"path": "user/mods",
			"restartRequired": false
		}
	],
	"exclusions": [
		// SPT Installer
		"BepInEx/plugins/spt",
		"BepInEx/patchers/spt-prepatch.dll",
		// Questing Bots
		"BepInEx/plugins/DanW-SPTQuestingBots/log",
		// Realism
		"user/mods/SPT-Realism/ProfileBackups",
		// Fika
		"user/mods/fika-server/types",
		"user/mods/fika-server/cache",
		"BepInEx/plugins/Fika.Dedicated.dll",
		// Live Flea Prices
		"user/mods/zzDrakiaXYZ-LiveFleaPrices/config",
		// Questing Bots
		"BepInEx/plugins/DanW-SPTQuestingBots/log",
		// EFTApi
		"BepInEx/plugins/kmyuhkyuk-EFTApi/cache",
		// Expanded Task Text (Accounts for bug with current version)
		"user/mods/ExpandedTaskText/src/**/cache.json",
		// Leaves Loot Fuckery
		"user/mods/leaves-loot_fuckery/output",
		// ADD MISSING QUEST WEAPON REQUIREMENTS
		"user/mods/zz_guiltyman-addmissingquestweaponrequirements/log.log",
		"user/mods/zz_guiltyman-addmissingquestweaponrequirements/user/logs",
		// Corter ModSync
		"BepInEx/patchers/Corter-ModSync-Patcher.dll",
		"**/*.nosync",
		"**/*.nosync.txt",
		// General server mods
		"user/mods/**/.git",
		"user/mods/**/node_modules",
		"user/mods/**/*.js",
		"user/mods/**/*.js.map",
		"**/*:Zone.Identifier"
	]
}`;

export class Config {
	private _globs: RegExp[];
	constructor(
		public syncPaths: Required<SyncPath>[],
		public exclusions: string[],
	) {
		this._globs = exclusions.map(glob);
	}

	public isExcluded(filePath: string): boolean {
		return this._globs.some(
			(glob) =>
				glob.test(unixPath(filePath)),
		);
	}
}
export class ConfigUtil {
	constructor(
		private vfs: VFS,
		private jsonUtil: JsonUtil,
		private modImporter: PreSptModLoader,
		private logger: ILogger,
	) { }

	/**
	 * @throws {Error} If the config file does not exist
	 */
	private async readConfigFile(): Promise<RawConfig> {
		const modPath = this.modImporter.getModPath("Corter-ModSync");
		const configPath = path.join(modPath, "config.jsonc");

		if (!this.vfs.exists(configPath))
			await this.vfs.writeFilePromisify(configPath, DEFAULT_CONFIG);

		return this.jsonUtil.deserializeJsonC(
			// @ts-expect-error I am right, SPT is wrong
			await this.vfs.readFilePromisify(configPath, { encoding: "utf-8" }),
			"config.jsonc",
		) as RawConfig;
	}

	/**
	 * @throws {Error} If the config is invalid
	 */
	private validateConfig(config: RawConfig): void {
		if (!Array.isArray(config.syncPaths))
			throw new Error(
				"Corter-ModSync: config.jsonc 'syncPaths' is not an array. Please verify your config is correct and try again.",
			);

		if (!Array.isArray(config.exclusions))
			throw new Error(
				"Corter-ModSync: config.jsonc 'exclusions' is not an array. Please verify your config is correct and try again.",
			);


		const uniquePaths = new Set();
		for (const syncPath of config.syncPaths) {
			if (typeof syncPath !== "string" && !("path" in syncPath))
				throw new Error(
					"Corter-ModSync: config.jsonc 'syncPaths' is missing 'path'. Please verify your config is correct and try again.",
				);

			if (
				typeof syncPath === "string"
					? path.isAbsolute(syncPath)
					: path.isAbsolute(syncPath.path)
			)
				throw new Error(
					`Corter-ModSync: SyncPaths must be relative to SPT server root. Invalid path '${syncPath}'`,
				);

			if (
				path
					.relative(
						process.cwd(),
						path.resolve(
							process.cwd(),
							typeof syncPath === "string" ? syncPath : syncPath.path,
						),
					)
					.startsWith("..")
			)
				throw new Error(
					`Corter-ModSync: SyncPaths must within SPT server root. Invalid path '${syncPath}'`,
				);

			if (uniquePaths.has(typeof syncPath === "string" ? syncPath : syncPath.path))
				throw new Error(
					`Corter-ModSync: SyncPaths must be unique. Duplicate path '${syncPath}'`,
				)
		}
	}

	public async load(): Promise<Config> {
		const rawConfig = await this.readConfigFile();
		this.validateConfig(rawConfig);

		return new Config(
			[
				{
					enabled: true,
					enforced: true,
					silent: true,
					restartRequired: false,
					path: "ModSync.Updater.exe",
				},
				{
					enabled: true,
					enforced: true,
					silent: true,
					restartRequired: true,
					path: "BepInEx/plugins/Corter-ModSync.dll",
				},
				...rawConfig.syncPaths
					.map((syncPath) => ({
						enabled: true,
						enforced: false,
						silent: false,
						restartRequired: true,
						...(typeof syncPath === "string" ? { path: syncPath } : syncPath),
					}))
					.sort((a, b) => b.path.length - a.path.length),
			],
			rawConfig.exclusions,
		);
	}
}
