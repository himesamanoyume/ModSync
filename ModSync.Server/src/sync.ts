import path from "node:path";
import type { ILogger } from "@spt/models/spt/utils/ILogger";
import type { VFS } from "@spt/utils/VFS";
import type { Config, SyncPath } from "./config";
import { hashFile } from "./utility/imoHash";
import { HttpError, winPath } from "./utility/misc";
import { Semaphore } from "./utility/semaphore";

type ModFile = {
	hash: string;
	directory: boolean;
};

export class SyncUtil {
	private limiter = new Semaphore(1024);

	constructor(
		private vfs: VFS,
		private config: Config,
		private logger: ILogger,
	) {}

	private async getFilesInDir(baseDir: string, dir: string): Promise<string[]> {
		if (!this.vfs.exists(dir)) {
			this.logger.warning(
				`Corter-ModSync: Directory '${dir}' does not exist, will be ignored.`,
			);
			return [];
		}

		const stats = await this.vfs.statPromisify(dir);
		if (stats.isFile()) return [dir];

		const files: string[] = [];
		for (const fileName of this.vfs.getFiles(dir)) {
			const file = path.join(dir, fileName);

			if (this.config.isExcluded(file)) continue;

			files.push(file);
		}

		for (const dirName of this.vfs.getDirs(dir)) {
			const subDir = path.join(dir, dirName);

			if (this.config.isExcluded(subDir)) continue;

			const subFiles = await this.getFilesInDir(baseDir, subDir);
			if (
				this.vfs.getFiles(subDir).length === 0 &&
				this.vfs.getDirs(subDir).length === 0
			)
				files.push(subDir);

			files.push(...subFiles);
		}

		if (
			stats.isDirectory() &&
			this.vfs.getFiles(dir).length === 0 &&
			this.vfs.getDirs(dir).length === 0
		)
			files.push(dir);

		return files;
	}

	private async buildModFile(
		file: string,
		// biome-ignore lint/correctness/noEmptyPattern: <explanation>
		{}: Required<SyncPath>,
	): Promise<ModFile> {
		const stats = await this.vfs.statPromisify(file);
		if (stats.isDirectory()) return { hash: "", directory: true };

		let retryCount = 0;
		const lock = await this.limiter.acquire();
		while (true) {
			try {
				const hash = await hashFile(file);
				lock.release();

				return {
					hash,
					directory: false,
				};
			} catch (e) {
				if (
					e instanceof Error &&
					"code" in e &&
					e.code === "EBUSY" &&
					retryCount < 5
				) {
					this.logger.error(
						`Error reading '${file}'. Retrying (${retryCount}/5)...`,
					);
					await new Promise((resolve) => setTimeout(resolve, 500));
					retryCount++;
					continue;
				}

				this.logger.error(`Error reading '${file}'. Exiting...`);
				this.logger.error(`${e}`);
				throw new HttpError(
					500,
					`Corter-ModSync: Error reading '${file}'\n${e}`,
				);
			}
		}
	}

	public async hashModFiles(
		syncPaths: Config["syncPaths"],
	): Promise<Record<string, Record<string, ModFile>>> {
		const result: Record<string, Record<string, ModFile>> = {};
		const processedFiles = new Set<string>();

		const startTime = performance.now();
		for (const syncPath of syncPaths) {
			const files = await this.getFilesInDir(syncPath.path, syncPath.path);
			const filesResult: Record<string, ModFile> = {};

			for (const file of files) {
				if (processedFiles.has(winPath(file))) continue;

				filesResult[winPath(file)] = await this.buildModFile(file, syncPath);

				processedFiles.add(winPath(file));
			}

			result[winPath(syncPath.path)] = filesResult;
		}

		this.logger.info(
			`Corter-ModSync: Hashed ${processedFiles.size} files in ${performance.now() - startTime}ms`,
		);

		return result;
	}

	/**
	 * @throws {Error} If file path is invalid
	 */
	public sanitizeDownloadPath(
		file: string,
		syncPaths: Config["syncPaths"],
	): string {
		const normalized = path.join(
			path.normalize(file).replace(/^(\.\.(\/|\\|$))+/, ""),
		);

		for (const syncPath of syncPaths) {
			const fullPath = path.join(process.cwd(), syncPath.path);
			if (!path.relative(fullPath, normalized).startsWith("..")) {
				return normalized;
			}
		}

		throw new HttpError(
			400,
			`Corter-ModSync: Requested file '${file}' is not in an enabled sync path!`,
		);
	}
}
