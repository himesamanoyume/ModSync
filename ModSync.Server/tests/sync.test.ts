import { expect, describe, it, vi, beforeEach } from "vitest";
import { fs, vol } from "memfs";

import { SyncUtil } from "../src/sync";
import { Config } from "../src/config";
import { VFS } from "./utils/vfs";
import type { VFS as IVFS } from "@spt/utils/VFS";
import type { ILogger } from "@spt/models/spt/utils/ILogger";
import { mock } from "vitest-mock-extended";

vi.mock("node:fs", async () => {
	const { readFileSync } = await vi.importActual<typeof import("node:fs")>("node:fs");
	const { fs } = await vi.importActual<typeof import("memfs")>("memfs");

	return {
		...fs,
		readFileSync: vi.fn((path, options) => path.endsWith(".wasm") ? readFileSync("../ModSync.MetroHash/pkg/metrohash_bg.wasm") : fs.readFileSync(path, options)),
	}
});


describe("syncTests", async () => {
	const directoryStructure = {
		plugins: {
			"file1.dll": "Test content",
			"file2.dll": "Test content 2",
			"file2.dll.nosync": "",
			"file3.dll": "Test content 3",
			"file3.dll.nosync.txt": "",
			ModName: {
				IncludedSubdir: {
					"aFileToInclude": "Something",
				},
				"mod_name.dll": "Test content 4",
				".nosync": "",
			},
			OtherMod: {
				"other_mod.dll": "Test content 5",
				sounds: {},
				subdir: {
					"image.png": "Test Image",
					".nosync": "",
				},
			},
		},
		user: {
			mods: {
				testMod: {
					"test.ts": "Test content 6",
					"test.js": "Test content 6",
					"test.js.map": "Test content 7",
				},
			},
		},
	};

	const config = new Config(
		[
			{
				path: "plugins",
				enabled: true,
				enforced: false,
				silent: false,
				restartRequired: true,
			},
		],
		[
			"**/*.nosync",
			"**/*.nosync.txt",
			"plugins/file2.dll",
			"plugins/file3.dll",
			"plugins/ModName",
			"plugins/OtherMod/subdir",
			"user/mods/**/*.js",
			"user/mods/**/*.js.map",
		],
	);

	const vfs = new VFS();
	const logger = mock<ILogger>();
	const syncUtil = new SyncUtil(vfs as IVFS, config, logger);

	beforeEach(() => {
		vol.reset();
		vol.fromNestedJSON(directoryStructure);
	});

	describe("hashModFiles", () => {
		it("should hash mod files", async () => {
			const hashes = await syncUtil.hashModFiles(config.syncPaths);

			expect("plugins\\OtherMod\\other_mod.dll" in hashes.plugins).toBe(
				true,
			);

			expect(hashes).toMatchSnapshot();
		});

		it("should correctly hash multiple folders", async () => {
			const newConfig = new Config(
				[
					{
						path: "plugins",
						enabled: true,
						enforced: false,
						silent: false,
						restartRequired: true,
					},
				],
				config.exclusions,
			);

			const vfs = new VFS();
			const syncUtil = new SyncUtil(vfs as IVFS, newConfig, logger);

			const hashes = await syncUtil.hashModFiles(newConfig.syncPaths);

			expect(hashes).toMatchSnapshot();
		});

		it("previous syncpaths should override later ones", async () => {
			const newConfig = new Config(
				[
					{
						path: "plugins/OtherMod",
						enabled: true,
						enforced: false,
						silent: false,
						restartRequired: true,
					},
					{
						path: "plugins",
						enabled: true,
						enforced: false,
						silent: false,
						restartRequired: true,
					},
				],
				config.exclusions,
			);

			const syncUtil = new SyncUtil(vfs as IVFS, newConfig, logger);

			const hashes = await syncUtil.hashModFiles(newConfig.syncPaths);

			expect(hashes["plugins\\OtherMod"]).toHaveProperty("plugins\\OtherMod\\other_mod.dll");
			expect(hashes.plugins).not.toHaveProperty("plugins\\OtherMod\\other_mod.dll");
		});

		it("should correctly ignore folders that do not exist", async () => {
			const newConfig = new Config(
				[
					{
						path: "plugins",
						enabled: true,
						enforced: false,
						silent: false,
						restartRequired: true,
					},
					{
						path: "user/bananas",
						enabled: true,
						enforced: false,
						silent: false,
						restartRequired: false,
					},
				],
				config.exclusions,
			);

			const syncUtil = new SyncUtil(vfs as IVFS, newConfig, logger);

			const hashes = await syncUtil.hashModFiles(newConfig.syncPaths);

			expect(Object.keys(hashes.plugins)).toContain("plugins\\file1.dll");
			expect(Object.keys(hashes.plugins)).toContain(
				"plugins\\OtherMod\\other_mod.dll",
			);
			expect(logger.warning).toHaveBeenCalledWith(
				"Corter-ModSync: Directory 'user/bananas' does not exist, will be ignored.",
			);
		});

		it("should correctly hash folders that didn't exist initially but are created", async () => {
			const newConfig = new Config(
				[
					{
						path: "plugins",
						enabled: true,
						enforced: false,
						silent: false,
						restartRequired: true,
					},
					{
						path: "user/bananas",
						enabled: true,
						enforced: false,
						silent: false,
						restartRequired: false,
					},
				],
				config.exclusions,
			);

			const syncUtil = new SyncUtil(vfs as IVFS, newConfig, logger);

			const hashes = await syncUtil.hashModFiles(newConfig.syncPaths);

			expect(Object.keys(hashes.plugins)).toContain("plugins\\file1.dll");
			expect(Object.keys(hashes.plugins)).toContain(
				"plugins\\OtherMod\\other_mod.dll",
			);

			expect(logger.warning).toHaveBeenCalledWith(
				"Corter-ModSync: Directory 'user/bananas' does not exist, will be ignored.",
			);

			fs.mkdirSync("user/bananas", { recursive: true });
			fs.writeFileSync("user/bananas/test.txt", "test");

			const newHashes = await syncUtil.hashModFiles(newConfig.syncPaths);

			expect(newHashes).toMatchSnapshot();
		});

		it("should correctly include empty folders", async () => {
			const hashes = await syncUtil.hashModFiles(config.syncPaths);

			expect("plugins\\OtherMod\\sounds" in hashes.plugins).toBe(true);

			expect(hashes).toMatchSnapshot();
		});

		it("should correctly ignore folders with all files excluded", async () => {
			const newConfig = new Config(
				config.syncPaths,
				[
					...config.exclusions,
					"plugins/OtherMod/*",
				]
			);

			const syncUtil = new SyncUtil(vfs as IVFS, newConfig, logger);

			const hashes = await syncUtil.hashModFiles(newConfig.syncPaths);

			expect(hashes).toMatchSnapshot();
		});

		it("Should correctly hash a single file", async () => {
			const hashes = await syncUtil.hashModFiles([
				{
					enabled: true,
					enforced: false,
					path: "plugins/file1.dll",
					restartRequired: true,
					silent: false,
				},
			]);

			expect(hashes).toMatchSnapshot();
		});

		it("Should only hash a given file for one syncPath", async () => {
			const hashes = await syncUtil.hashModFiles([
				{
					enabled: true,
					enforced: false,
					path: "plugins/file1.dll",
					restartRequired: true,
					silent: false,
				},
				{
					enabled: true,
					enforced: false,
					path: "plugins/",
					restartRequired: true,
					silent: false,
				},
			]);

			expect(Object.values(hashes).flatMap(Object.keys).filter((path) => path === "plugins\\file1.dll").length).toBe(1);
			expect(hashes).toMatchSnapshot();
		});

		it("Exclusions can be overridden by more specific syncPaths", async () => {
			const newConfig = new Config(
				[
					...config.syncPaths,
					{
						path: "plugins/ModName/IncludedSubdir",
						enabled: true,
						enforced: false,
						silent: false,
						restartRequired: true,
					}
				],
				config.exclusions,
			)

			const syncUtil = new SyncUtil(vfs as IVFS, newConfig, logger);
			const hashes = await syncUtil.hashModFiles(newConfig.syncPaths);


			expect(hashes["plugins\\ModName\\IncludedSubdir"]).toHaveProperty("plugins\\ModName\\IncludedSubdir\\aFileToInclude");
			expect(hashes).toMatchSnapshot();
		});
	});

	describe("sanitizeDownloadPath", () => {
		it("should sanitize correct download paths", () => {
			expect(
				syncUtil.sanitizeDownloadPath("plugins\\file1.dll", config.syncPaths),
			).toBe("plugins\\file1.dll");
		});

		it("should throw for download paths outside SPT root", () => {
			expect(() => {
				syncUtil.sanitizeDownloadPath(
					"plugins\\..\\file1.dll",
					config.syncPaths,
				);
			}).toThrow();
		});

		it("should throw for files not in syncPath", () => {
			expect(() => {
				syncUtil.sanitizeDownloadPath("otherDir\\file1.dll", config.syncPaths);
			}).toThrow();
		});
	});
});
