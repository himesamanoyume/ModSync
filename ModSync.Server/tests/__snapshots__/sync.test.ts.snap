// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`syncTests > hashModFiles > Exclusions can be overridden by more specific syncPaths 1`] = `
{
  "plugins": {
    "plugins\\OtherMod\\other_mod.dll": {
      "directory": false,
      "hash": "0e1546a7aa36a802e102888baa1681fc",
    },
    "plugins\\OtherMod\\sounds": {
      "directory": true,
      "hash": "",
    },
    "plugins\\file1.dll": {
      "directory": false,
      "hash": "0ce304b7ff04260d67adfdee0af9dd3b",
    },
  },
  "plugins\\ModName\\IncludedSubdir": {
    "plugins\\ModName\\IncludedSubdir\\aFileToInclude": {
      "directory": false,
      "hash": "09d2277efc596b918a7415af7ea34d50",
    },
  },
}
`;

exports[`syncTests > hashModFiles > Should correctly hash a single file 1`] = `
{
  "plugins\\file1.dll": {
    "plugins\\file1.dll": {
      "directory": false,
      "hash": "0ce304b7ff04260d67adfdee0af9dd3b",
    },
  },
}
`;

exports[`syncTests > hashModFiles > Should only hash a given file for one syncPath 1`] = `
{
  "plugins\\": {
    "plugins\\OtherMod\\other_mod.dll": {
      "directory": false,
      "hash": "0e1546a7aa36a802e102888baa1681fc",
    },
    "plugins\\OtherMod\\sounds": {
      "directory": true,
      "hash": "",
    },
  },
  "plugins\\file1.dll": {
    "plugins\\file1.dll": {
      "directory": false,
      "hash": "0ce304b7ff04260d67adfdee0af9dd3b",
    },
  },
}
`;

exports[`syncTests > hashModFiles > should correctly hash folders that didn't exist initially but are created 1`] = `
{
  "plugins": {
    "plugins\\OtherMod\\other_mod.dll": {
      "directory": false,
      "hash": "0e1546a7aa36a802e102888baa1681fc",
    },
    "plugins\\OtherMod\\sounds": {
      "directory": true,
      "hash": "",
    },
    "plugins\\file1.dll": {
      "directory": false,
      "hash": "0ce304b7ff04260d67adfdee0af9dd3b",
    },
  },
  "user\\bananas": {
    "user\\bananas\\test.txt": {
      "directory": false,
      "hash": "047d7c20c1ea6d66718ac45146ab06cd",
    },
  },
}
`;

exports[`syncTests > hashModFiles > should correctly hash multiple folders 1`] = `
{
  "plugins": {
    "plugins\\OtherMod\\other_mod.dll": {
      "directory": false,
      "hash": "0e1546a7aa36a802e102888baa1681fc",
    },
    "plugins\\OtherMod\\sounds": {
      "directory": true,
      "hash": "",
    },
    "plugins\\file1.dll": {
      "directory": false,
      "hash": "0ce304b7ff04260d67adfdee0af9dd3b",
    },
  },
  "user\\mods": {
    "user\\mods\\testMod\\test.ts": {
      "directory": false,
      "hash": "0ed5f429912dfd9fb77009353ab0f889",
    },
  },
}
`;

exports[`syncTests > hashModFiles > should correctly ignore folders with all files excluded 1`] = `
{
  "plugins": {
    "plugins\\file1.dll": {
      "directory": false,
      "hash": "0ce304b7ff04260d67adfdee0af9dd3b",
    },
  },
}
`;

exports[`syncTests > hashModFiles > should correctly include empty folders 1`] = `
{
  "plugins": {
    "plugins\\OtherMod\\other_mod.dll": {
      "directory": false,
      "hash": "0e1546a7aa36a802e102888baa1681fc",
    },
    "plugins\\OtherMod\\sounds": {
      "directory": true,
      "hash": "",
    },
    "plugins\\file1.dll": {
      "directory": false,
      "hash": "0ce304b7ff04260d67adfdee0af9dd3b",
    },
  },
}
`;

exports[`syncTests > hashModFiles > should hash mod files 1`] = `
{
  "plugins": {
    "plugins\\OtherMod\\other_mod.dll": {
      "directory": false,
      "hash": "0e1546a7aa36a802e102888baa1681fc",
    },
    "plugins\\OtherMod\\sounds": {
      "directory": true,
      "hash": "",
    },
    "plugins\\file1.dll": {
      "directory": false,
      "hash": "0ce304b7ff04260d67adfdee0af9dd3b",
    },
  },
}
`;
